File: Sources/FWObjC/Module/Controller/WebViewController.m
//
//  WebViewController.m
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

#import "WebViewController.h"
#import "NavigationController.h"
#import <objc/runtime.h>

#if FWMacroSPM

@interface UIView ()

- (NSArray<NSLayoutConstraint *> *)fw_pinEdgesToSuperview:(UIEdgeInsets)insets;

@end

@interface NSObject ()

- (NSString *)fw_observeProperty:(NSString *)property block:(void (^)(id object, NSDictionary<NSKeyValueChangeKey, id> *change))block;

@end

@interface UIViewController ()

@property (nonatomic, copy, nullable) BOOL (^fw_allowsPopGesture)(void);

@end

@interface UIBarButtonItem ()

+ (instancetype)fw_itemWithObject:(nullable id)object block:(nullable void (^)(id sender))block;

@end

#else

#import <FWFramework/FWFramework-Swift.h>

#endif

#pragma mark - __FWViewControllerManager+__FWWebViewController

@implementation __FWViewControllerManager (__FWWebViewController)

+ (void)load
{
    __FWViewControllerIntercepter *intercepter = [[__FWViewControllerIntercepter alloc] init];
    intercepter.viewDidLoadIntercepter = @selector(webViewControllerViewDidLoad:);
    intercepter.forwardSelectors = @{
        @"webView" : @"fw_innerWebView",
        @"webItems" : @"fw_innerWebItems",
        @"setWebItems:" : @"fw_innerSetWebItems:",
        @"webRequest" : @"fw_innerWebRequest",
        @"setWebRequest:" : @"fw_innerSetWebRequest:",
        @"setupWebLayout" : @"fw_innerSetupWebLayout",
    };
    [[__FWViewControllerManager sharedInstance] registerProtocol:@protocol(__FWWebViewController) withIntercepter:intercepter];
}

- (void)webViewControllerViewDidLoad:(UIViewController<__FWWebViewController> *)viewController
{
    __FWWebView *webView = [viewController webView];
    webView.delegate = viewController;
    [viewController.view addSubview:webView];
    
    __weak __typeof(viewController) weakController = viewController;
    [webView fw_observeProperty:@"title" block:^(WKWebView *webView, NSDictionary *change) {
        weakController.navigationItem.title = webView.title;
    }];
    viewController.fw_allowsPopGesture = ^BOOL{
        return !weakController.webView.canGoBack;
    };
    
    if (self.hookWebViewController) {
        self.hookWebViewController(viewController);
    }
    
    if ([viewController respondsToSelector:@selector(setupWebView)]) {
        [viewController setupWebView];
    }
    
    [viewController setupWebLayout];
    [webView setNeedsLayout];
    [webView layoutIfNeeded];
    
    if ([viewController respondsToSelector:@selector(setupWebBridge:)]) {
        id<WKNavigationDelegate> delegate = webView.navigationDelegate;
        __FWWebViewJsBridge *bridge = [__FWWebViewJsBridge bridgeForWebView:webView];
        [bridge setWebViewDelegate:delegate];
        webView.fw_jsBridge = bridge;
        
        [viewController setupWebBridge:bridge];
    }
    
    NSArray *webItems = viewController.webItems;
    if (webItems.count < 1 || !viewController.navigationController) {
        viewController.webView.webRequest = viewController.webRequest;
        return;
    }
    
    NSMutableArray<UIBarButtonItem *> *leftItems = [NSMutableArray array];
    for (int i = 0; i < webItems.count; i++) {
        id webItem = webItems[i];
        if ([webItem isKindOfClass:[UIBarButtonItem class]]) {
            [leftItems addObject:webItem];
        } else {
            if (i == 0) {
                UIBarButtonItem *leftItem = [UIBarButtonItem fw_itemWithObject:webItem block:^(id sender) {
                    if (weakController.webView.canGoBack) {
                        [weakController.webView goBack];
                    } else {
                        if (weakController.navigationController) {
                            if ([weakController.navigationController popViewControllerAnimated:YES]) return;
                        }
                        if (weakController.presentingViewController) {
                            [weakController dismissViewControllerAnimated:YES completion:nil];
                            return;
                        }
                        
                        WKBackForwardListItem *firstItem = weakController.webView.backForwardList.backList.firstObject;
                        if (firstItem != nil) {
                            [weakController.webView goToBackForwardListItem:firstItem];
                        }
                    }
                }];
                [leftItems addObject:leftItem];
            } else {
                UIBarButtonItem *leftItem = [UIBarButtonItem fw_itemWithObject:webItem block:^(id sender) {
                    if (weakController.navigationController) {
                        if ([weakController.navigationController popViewControllerAnimated:YES]) return;
                    }
                    if (weakController.presentingViewController) {
                        [weakController dismissViewControllerAnimated:YES completion:nil];
                        return;
                    }
                    
                    WKBackForwardListItem *firstItem = weakController.webView.backForwardList.backList.firstObject;
                    if (firstItem != nil) {
                        [weakController.webView goToBackForwardListItem:firstItem];
                    }
                }];
                [leftItems addObject:leftItem];
            }
        }
    }
    
    BOOL showClose = YES;
    if (viewController.navigationController.viewControllers.firstObject == viewController &&
        viewController.navigationController.presentingViewController.presentedViewController != viewController.navigationController) {
        showClose = NO;
    }
    viewController.navigationItem.leftBarButtonItems = [NSArray arrayWithObjects:showClose ? leftItems.firstObject : nil, nil];
    [viewController.webView fw_observeProperty:@"canGoBack" block:^(WKWebView *webView, NSDictionary *change) {
        if (webView.canGoBack) {
            weakController.navigationItem.leftBarButtonItems = [leftItems copy];
        } else {
            weakController.navigationItem.leftBarButtonItems = [NSArray arrayWithObjects:showClose ? leftItems.firstObject : nil, nil];
        }
    }];
    
    viewController.webView.webRequest = viewController.webRequest;
}

@end

#pragma mark - UIViewController+__FWWebViewController

@interface UIViewController (__FWWebViewController)

@end

@implementation UIViewController (__FWWebViewController)

- (__FWWebView *)fw_innerWebView
{
    __FWWebView *webView = objc_getAssociatedObject(self, _cmd);
    if (!webView) {
        if ([self respondsToSelector:@selector(setupWebConfiguration)]) {
            webView = [[__FWWebView alloc] initWithFrame:CGRectZero configuration:[(id<__FWWebViewController>)self setupWebConfiguration]];
        } else {
            webView = [[__FWWebView alloc] initWithFrame:CGRectZero];
        }
        objc_setAssociatedObject(self, _cmd, webView, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    }
    return webView;
}

- (NSArray *)fw_innerWebItems
{
    return objc_getAssociatedObject(self, @selector(fw_innerWebItems));
}

- (void)fw_innerSetWebItems:(NSArray *)webItems
{
    objc_setAssociatedObject(self, @selector(fw_innerWebItems), webItems, OBJC_ASSOCIATION_COPY_NONATOMIC);
}

- (id)fw_innerWebRequest
{
    return objc_getAssociatedObject(self, @selector(fw_innerWebRequest));
}

- (void)fw_innerSetWebRequest:(id)webRequest
{
    objc_setAssociatedObject(self, @selector(fw_innerWebRequest), webRequest, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    
    if (self.isViewLoaded) {
        __FWWebView *webView = [(id<__FWWebViewController>)self webView];
        webView.webRequest = webRequest;
    }
}

- (void)fw_innerSetupWebLayout
{
    __FWWebView *webView = [(id<__FWWebViewController>)self webView];
    [webView fw_pinEdgesToSuperview:UIEdgeInsetsZero];
}

@end

File: Sources/FWFramework/Module/Controller/WebViewController.swift
//
//  WebViewController.swift
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

import UIKit
import WebKit

// MARK: - WebViewControllerProtocol
/// 网页视图控制器协议，可覆写
///
/// WebViewControllerProtocol默认未开启WebView重用，如需开启，方式如下：
/// 1. 只需配置WebView.reuseConfigurationBlock并设置ViewControllerManager.webViewReuseIdentifier不为nil即可
/// 2. 其他初始化、预加载、回收等重用操作框架会自动处理，详见源码
/// 3. 如果需要预缓存资源，配置WebView.reusePreloadUrlBlock后再设置webViewReuseIdentifier即可
public protocol WebViewControllerProtocol: ViewControllerProtocol, WebViewDelegate {
    
    /// 网页视图，默认显示滚动条，启用前进后退手势
    var webView: WebView { get }

    /// 网页请求，设置后会自动加载，支持NSString|NSURL|NSURLRequest。默认nil
    var webRequest: Any? { get set }

    /// 渲染网页配置，setupWebView之前调用，默认空实现
    func setupWebConfiguration(_ configuration: WKWebViewConfiguration)

    /// 渲染网页视图，setupSubviews之前调用，默认空实现
    func setupWebView()

    /// 渲染网页视图布局，setupSubviews之前调用，默认铺满
    func setupWebLayout()

    /// 渲染网页桥接，jsBridgeEnabled启用后生效，setupSubviews之前调用，默认空实现
    func setupWebBridge(_ bridge: WebViewJSBridge)
    
}

extension WebViewControllerProtocol where Self: UIViewController {
    
    /// 网页视图，默认显示滚动条，启用前进后退手势
    public var webView: WebView {
        if let result = fw.property(forName: "webView") as? WebView {
            return result
        } else {
            var result: WebView
            if let reuseIdentifier = ViewControllerManager.shared.webViewReuseIdentifier {
                result = ReusableViewPool.shared.dequeueReusableView(with: WebView.self, viewHolder: self, reuseIdentifier: reuseIdentifier)
            } else {
                let configuration = WKWebView.fw_defaultConfiguration()
                setupWebConfiguration(configuration)
                result = WebView(frame: .zero, configuration: configuration)
            }
            fw.setProperty(result, forName: "webView")
            return result
        }
    }
    
    /// 网页请求，设置后会自动加载，支持NSString|NSURL|NSURLRequest。默认nil
    public var webRequest: Any? {
        get {
            return fw.property(forName: "webRequest")
        }
        set {
            fw.setProperty(newValue, forName: "webRequest")
            
            if isViewLoaded {
                webView.webRequest = newValue
            }
        }
    }
    
    /// 渲染网页配置，setupWebView之前调用，默认空实现
    public func setupWebConfiguration(_ configuration: WKWebViewConfiguration) {}

    /// 渲染网页视图，setupSubviews之前调用，默认空实现
    public func setupWebView() {}
    
    /// 渲染网页视图布局，setupSubviews之前调用，默认铺满
    public func setupWebLayout() {
        webView.fw.pinEdges()
    }
    
    /// 渲染网页桥接，setupSubviews之前调用，默认空实现
    public func setupWebBridge(_ bridge: WebViewJSBridge) {}
    
}

// MARK: - ViewControllerManager+WebViewControllerProtocol
internal extension ViewControllerManager {
    
    func webViewControllerViewDidLoad(_ viewController: UIViewController) {
        guard let viewController = viewController as? UIViewController & WebViewControllerProtocol else { return }
        
        let webView = viewController.webView
        webView.delegate = viewController
        viewController.view.addSubview(webView)
        
        if webViewReuseIdentifier != nil {
            viewController.fw_observeLifecycleState(object: webView) { [weak self] _, state, webView in
                guard self?.webViewReuseIdentifier != nil,
                      state == .didDeinit,
                      let webView = webView as? WebView else { return }
                
                ReusableViewPool.shared.recycleReusableView(webView)
            }
        }
        
        webView.fw.observeProperty(\.title) { [weak viewController] _, _ in
            viewController?.navigationItem.title = viewController?.webView.title
        }
        viewController.fw_allowsPopGesture = { [weak viewController] in
            return !(viewController?.webView.canGoBack ?? false)
        }
        
        hookWebViewController?(viewController)
        
        viewController.setupWebView()
        viewController.setupWebLayout()
        webView.setNeedsLayout()
        webView.layoutIfNeeded()
        
        if webView.fw_jsBridgeEnabled, let bridge = webView.fw_setupJsBridge() {
            viewController.setupWebBridge(bridge)
        }
        
        if webView.fw_navigationItems != nil {
            webView.fw_setupNavigationItems(viewController)
        }
        
        webView.webRequest = viewController.webRequest
    }
    
}
